### PR TITLE
xds: Add counter and gauge metrics 

### PR DESCRIPTION
**Pull request by DNVindhya**
This PR implements xDS client defined in [A78](https://github.com/grpc/proposal/blob/master/A78-grpc-metrics-wrr-pf-xds.md#xdsclient).

Counters
- grpc.xds_client.server_failure
- grpc.xds_client.resource_updates_valid
- grpc.xds_client.resource_updates_invalid

Gauges
- grpc.xds_client.connected
- grpc.xds_client.resources

The `grpc.xds.authority` label is missing, and will be added in a later PR.

### CODE REVIEW

**Review by ejona86 — COMMENTED** [review]
Sending what I have.

File: xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java:134]: Run the two concurrently (start both, and then get() both), so we don't pay any latencies twice?

File: xds/src/test/java/io/grpc/xds/XdsClientMetricReporterImplTest.java
Comment (ejona86) [inline xds/src/test/java/io/grpc/xds/XdsClientMetricReporterImplTest.java:143]: Never, ever use a spy. The way they are implemented essentially invalidates any testing using them. Use a mock with `delegatesTo(realImpl)`.

File: xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java
Comment (DNVindhya) [reply xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java]: Makes sense, started both.

File: xds/src/test/java/io/grpc/xds/XdsClientMetricReporterImplTest.java
Comment (DNVindhya) [reply xds/src/test/java/io/grpc/xds/XdsClientMetricReporterImplTest.java]: When I was modifying the test to remove spy, realized I do not need `delegatesTo` either.

File: xds/src/main/java/io/grpc/xds/XdsClientMetricReporter.java
Comment (DNVindhya) [reply xds/src/main/java/io/grpc/xds/XdsClientMetricReporter.java]: Split the interface into two different interfaces. PTAL

File: xds/src/test/java/io/grpc/xds/XdsClientMetricReporterImplTest.java
Comment (larry-safran) [inline xds/src/test/java/io/grpc/xds/XdsClientMetricReporterImplTest.java:165]: You don't need to put the class name in anymore (it used to be a problem well before I was on the team).  It's easier to read without it and doesn't need to be changed when the argument changes.

File: xds/src/main/java/io/grpc/xds/SharedXdsClientPoolProvider.java
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/SharedXdsClientPoolProvider.java:124]: This isn't providing any value as `non-static`. Either we should move it up to the provider so it can remain `static` or construct the object when creating XdsClientImpl.

(Actually, we could make Provider a singleton in ExponentialBackoffPolicy itself. But the object is small, so it doesn't matter much.)

File: xds/src/main/java/io/grpc/xds/XdsNameResolver.java
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/XdsNameResolver.java:131]: Why is this volatile? It is set once during construction. Seems it could even be `final`.

File: xds/src/main/java/io/grpc/xds/XdsServerWrapper.java
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/XdsServerWrapper.java:175]: Should there be a TODO here?

File: xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java:554]: The point of using a callback is mostly to avoid allocating the map. If you're going to construct the map, you can instead just "return" it as the `future` result. See my other comment in getResourceCountsByType().

Although, I'd actually consider getting rid of this method and using `getSubscribedResourcesMetadataSnapshot()` instead. It looks like it has access to all the data needed here and avoids polluting XdsClientImpl with the very specific form of data you need. In particular, "nacked_but_cached" seems highly specific to this one consumer of the data.

(You can contrast this method with `reportServerConnections()`, which works well as a callback as it just loops over the existing data. Although the data it returns is also small, so it could have been fine to copy it.)

File: xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java:568]: This map is only written to in this method, so it isn't actually necessary. If `ResourceCallback` was an argument to this method:
```java
// instead of:
      resourceCountsByType.put(resourceType, resourceCountsByState);
// you could do:
      resourceCountsByState.forEach((cacheState, count) ->
          callback.reportResourceCountGauge(count, cacheState, resourceType.typeUrl(), target));
```

There's not many XdsResourceType nor cache_state, so creating the full map isn't that much memory. Most of the cost right now is `toLowerCase()` and `Long` object creation for each entry. And optimizing those has no real impact to the memory cost of the data structure you return.

Although it looks like this is going to have to be reorganized when `authority` is added, that still won't add much additional memory.

But to really benefit from the callback, we probably shouldn't do any aggregation within XdsClientImpl. Instead, call the callback with the raw data and have it aggregate as it needs. That means passing `ResourceMetadataStatus` and `isResourceCached` separately and letting it determine the string to use. That makes the implementation here less sensitive to the whims of the metric reporter.

File: xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java:116]: Don't pass `target` into XdsClientImpl. It isn't using it; it is just echoing it right back out. Instead, `target` should be remembered by `XdsClientMetricReporterImpl` and "injected" while reporting.

File: xds/src/main/java/io/grpc/xds/SharedXdsClientPoolProvider.java
Comment (DNVindhya) [reply xds/src/main/java/io/grpc/xds/SharedXdsClientPoolProvider.java]: Moved it up to the provider.

File: xds/src/main/java/io/grpc/xds/XdsNameResolver.java
Comment (DNVindhya) [reply xds/src/main/java/io/grpc/xds/XdsNameResolver.java]: That was an oversight, updated.

File: xds/src/main/java/io/grpc/xds/XdsServerWrapper.java
Comment (DNVindhya) [reply xds/src/main/java/io/grpc/xds/XdsServerWrapper.java:177]: Added

File: xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java
Comment (DNVindhya) [reply xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java]: Done

File: xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java
Comment (DNVindhya) [reply xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java]: `getSubscribedResourcesMetadataSnapshot()` is missing `subscriber.data` which tells us if resource is cached and used in determining `grpc.xds.cache_state` attribute value. 
As discussed, I will try adding `subcriber.data` value as a new field in ResourceMetadata and after that I will move the processing / formatting of values to `XdsClientMetricReporterImpl`.

File: xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java
Comment (DNVindhya) [reply xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java]: I initially separated to into multiple methods for testing the logic of aggregating resource counts separately, but instead tested using callback itself. Removed `getResourceCountsByType()` and `reportResourceCountsToCallback()`. 

> Instead, call the callback with the raw data and have it aggregate as it needs.
Will make this change after populating `subscriber.data` value in `ResourceMetadata`.

File: xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java
Comment (ejona86) [reply xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java]: You might just add a `boolean` for whether there is data cached. That way it does stay "metadata."

File: xds/src/main/java/io/grpc/xds/SharedXdsClientPoolProvider.java
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/SharedXdsClientPoolProvider.java:124]: `protected` is weird to see here, as `SharedXdsClientPoolProvider` is `final`. Revert back to package-private? (It doesn't actually change anything, but just confusing.)

File: xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java:44]: Why is this public? Make it package-private?

File: xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java:127]: Since the reporter is reused, let's also set `gaugeRegistration = null` just after this, so it'll be a clearer when debugging.

File: xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java:120]: Since you're making an anonymous class anyway, you can also pass XdsClient here. Less state.

File: xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java:116]: Since the reporter is reused, I suggest at least adding `assert gaugeRegistration == null` to catch lifetime bugs. (Which would require setting it to null in close().)

File: xds/src/main/java/io/grpc/xds/client/ControlPlaneClient.java
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/client/ControlPlaneClient.java:236]: I think this method should just be `return !streamClosedNoResponse`. `shutdown` shouldn't matter, since this method should not be called when shutdown. `if (adsStream == null) return true` is the same as `if (adsStream == null) return !streamClosedNoResponse` because of earlier checks. And `responseReceived == !streamClosedNoResponse` when `adsStream.closed == true`. When `adsStream.closed == false`, it seems `!streamClosedNoResponse` is more correct than `responseReceived`, as we don't want to incorrectly say it is working.

File: xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java
Comment (DNVindhya) [reply xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java]: Moved resource count by cache state metric aggregating and reporting to `XdsClientmetricReporterImpl". Added a boolean `isCached` to `ResourceMetadata`. PTAL

**Review by larry-safran — APPROVED** [review]

File: xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java
Comment (DNVindhya) [reply xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java:397]: Counter of xDS servers going from healthy to unhealthy  metric (`grpc.xds_client.server_failure`) is reported as part of `handleStreamClosed` and the metric needs `grpc.xds.server` as an attribute value. 
Now with the help of anonymous  class we are able to create a `XdsResponseHandler` instance for every ControlPlaneClient and provide ControlPlaneClient's `ServerInfo` to `XdsClientImpl.this.handleStreamClosed`.

File: xds/src/main/java/io/grpc/xds/client/XdsClient.java
Comment (DNVindhya) [reply xds/src/main/java/io/grpc/xds/client/XdsClient.java]: Added quotes around working.

File: xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java:168]: Delete? The future doesn't seem to be doing anything because this runs in-line.

File: xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java:189]: Instead of toLower, which is going to allocate every item, can we just have a `switch` here that returns string literals?

File: xds/src/main/java/io/grpc/xds/client/XdsClient.java
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/client/XdsClient.java:406]: The return type should be `ListenableFuture<Void>` (or `Future<Void>`, if we don't want/care about notifications). `SettableFuture` should always be an implementation detail.

**Review by ejona86 — APPROVED** [review]

File: xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java
Comment (DNVindhya) [reply xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java]: Removed

File: xds/src/main/java/io/grpc/xds/client/XdsClient.java
Comment (DNVindhya) [reply xds/src/main/java/io/grpc/xds/client/XdsClient.java]: TIL, updated to return `Future<Void>` instead.

File: xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java:203]: private or package-private? Or just inline it? It doesn't get any benefit from being here instead of the parent class.

Comment (DNVindhya) [reply xds/src/main/java/io/grpc/xds/XdsClientMetricReporterImpl.java]: Updated to package-private.

**Comment by DNVindhya** [comment]
@ejona86 and @larry-safran thanks for both of your approvals. I have addressed all your comments. If you don't have any new/pending comments can I go ahead and merge?

**Comment by larry-safran** [comment]
Once you have approvals from everyone, you can go ahead a merge.

**Pull request merged**
